### PR TITLE
stm32 entropy fix

### LINKED ISSUE #66312: tests: crypto: rand32 test issue regression
Describe the bug:

in our bench tester many board test fail now  with:

tests/crypto/rand32
for example:

west build -p always -b   nucleo_f746zg  tests/crypto/rand32
image

test is blocked

it is due to this commit:
https://github.com/zephyrproject-rtos/zephyr/commit/78af1c1e3878c410067bd49846771300fa86b840

entropy: stm32: suspend when pool is full
78af1c1e3878c410067bd49846771300fa86b840
author: Dawid Niedzwiecki <dawidn@google.com>  2023-12-01 11:27:37
Committer: Maureen Helm <maureen.helm@analog.com>  2023-12-05 23:27:53

thanks  to have a look

### PR DESCRIPTION
**Pull request by niedzwiecki-dawid**
Enable the RNG module for the get_entropy_isr function if needed.

Fixes #66312

### CODE REVIEW

File: drivers/entropy/entropy_stm32.c [hunk 1]
```diff
@@ -595,13 +597,18 @@ static int entropy_stm32_rng_get_entropy_isr(const struct device *dev,
...
 		irq_enabled = irq_is_enabled(IRQN);
 		irq_disable(IRQN);
+		rng_enabled = LL_RNG_IsEnabled(rng);
 		irq_unlock(key);
 
+		if (!rng_enabled) {
+			entropy_stm32_resume(dev);
+		}
 		rng_already_acquired = z_stm32_hsem_is_owned(CFG_HW_RNG_SEMID);
 		acquire_rng();
```
[the remark this answers is not in the document]
Comment (niedzwiecki-dawid) [reply drivers/entropy/entropy_stm32.c:609]: It should work. Looks cleaner now :-)

**Comment by erwango** [comment]
Found some remaining failing tests with current solution:
```
   tests/crypto/rand32/crypto.rand32, nucleo_f091rc (Regression)
   tests/crypto/rand32/crypto.rand32, nucleo_h743zi
   tests/crypto/rand32/crypto.rand32.random_ctr_drbg, nucleo_h743zi
   tests/crypto/rand32/crypto.rand32.random_hw_xoshiro, nucleo_h743zi
   tests/crypto/rand32/crypto.rand32.random_sw_systimer, nucleo_f091rc (Regression)
   tests/crypto/rand32/crypto.rand32.random_sw_systimer, nucleo_h743zi
   tests/crypto/tinycrypt/crypto.tinycrypt, nucleo_f091rc (Regression)
   tests/crypto/tinycrypt/crypto.tinycrypt, nucleo_h743zi
   tests/kernel/mem_protect/stack_random/kernel.memory_protection.stack_random, nucleo_f091rc (Regression)
   tests/kernel/mem_protect/stack_random/kernel.memory_protection.stack_random, nucleo_h743zi
   tests/kernel/mem_protect/stackprot/kernel.memory_protection.stackprot, nucleo_f091rc (Regression)
   tests/kernel/mem_protect/sys_sem/kernel.memory_protection.sys_sem.nouser, nucleo_f091rc (Regression)
   tests/net/lib/dns_resolve/net.dns.resolve, nucleo_h743zi
   tests/net/lib/dns_resolve/net.dns.resolve.no_ipv6, nucleo_h743zi
   tests/net/lib/dns_resolve/net.dns.resolve.preempt, nucleo_h743zi
```
Tested with `twister --ninja --runtime-artifact-cleanup -e logging -e cmsis_dsp -T tests/crypto -T tests/kernel/mem_protect -T tests/net/lib/dns_resolve/`
Boards are the ones from our test farm (you can see the list here: https://github.com/zephyrproject-rtos/test_results/pull/1788/files)

**Comment by niedzwiecki-dawid** [comment]
> tests/kernel/mem_protect/stackprot/kernel.memory_protection.stackprot, nucleo_f091rc (Regression)
>    tests/kernel/mem_protect/sys_sem/kernel.memory_protection.sys_sem.nouser, nucleo_f091rc (Regression)

Can I somehow run tests on these boards or at least see tests logs?

**Comment by erwango** [comment]
> Can I somehow run tests on these boards or at least see tests logs?

I'm running a new test session. Note that this target is also impacted by https://github.com/zephyrproject-rtos/zephyr/issues/66388 and additionally was maybe in a bad state. I'll send new results when received

**Comment by niedzwiecki-dawid** [comment]
Ok, thanks. I've ordered nucleo_h743zi anyway and should be able to run tests locally in 1/2 days.

**Comment by erwango** [comment]
New session results:
```
   tests/crypto/rand32/crypto.rand32, nucleo_h743zi
   tests/crypto/rand32/crypto.rand32.random_ctr_drbg, nucleo_h743zi
   tests/crypto/rand32/crypto.rand32.random_hw_xoshiro, nucleo_h743zi
   tests/crypto/rand32/crypto.rand32.random_sw_systimer, nucleo_h743zi
   tests/crypto/tinycrypt/crypto.tinycrypt, nucleo_h743zi
   tests/kernel/mem_protect/stack_random/kernel.memory_protection.stack_random, nucleo_h743zi
   tests/net/lib/dns_resolve/net.dns.resolve, nucleo_h743zi
   tests/net/lib/dns_resolve/net.dns.resolve.no_ipv6, nucleo_h743zi
   tests/net/lib/dns_resolve/net.dns.resolve.preempt, nucleo_h743zi
 ```

**Comment by erwango** [comment]
Back trace when running `tests/crypto/rand32` on a H7 board:
```
generate_from_isr (len=1, buf=0x2400098c <ztest_thread_stack+972> "") at /local/mcu/zephyrproject/zephyr/drivers/entropy/entropy_stm32.c:364
364				__SEV();
(gdb) bt
#0  generate_from_isr (len=1, buf=0x2400098c <ztest_thread_stack+972> "")
    at /local/mcu/zephyrproject/zephyr/drivers/entropy/entropy_stm32.c:364
#1  entropy_stm32_rng_get_entropy_isr (dev=<optimized out>, 
    buf=0x2400098c <ztest_thread_stack+972> "", len=<optimized out>, flags=<optimized out>)
    at /local/mcu/zephyrproject/zephyr/drivers/entropy/entropy_stm32.c:608
#2  0x0800385a in entropy_get_entropy_isr (dev=<optimized out>, flags=1, length=1, 
    buffer=0x2400098c <ztest_thread_stack+972> "")
    at /local/mcu/zephyrproject/zephyr/include/zephyr/drivers/entropy.h:116
#3  z_early_rand_get (buf=buf@entry=0x2400098c <ztest_thread_stack+972> "", length=length@entry=1)
    at /local/mcu/zephyrproject/zephyr/kernel/init.c:517
#4  0x0800060c in rand32_common_test_rand32 ()
```

**Comment by niedzwiecki-dawid** [comment]
> Member

Ok, thanks. So the problem occurs only on H7, right?

**Comment by erwango** [comment]
> Ok, thanks. So the problem occurs only on H7, right?

Indeed

**Comment by niedzwiecki-dawid** [comment]
I have to run the tests on device. I'm not able to find the issue without actual debug.

**Comment by niedzwiecki-dawid** [comment]
@erwango Would you be able to retest with the current version?

File: drivers/entropy/entropy_stm32.c [hunk 2]
```diff
@@ -143,6 +145,7 @@ static int entropy_stm32_resume(const struct device *dev)
 	res = clock_control_on(dev_data->clock,
 			(clock_control_subsys_t)&dev_cfg->pclken[0]);
 	LL_RNG_Enable(rng);
+	LL_RNG_EnableIT(rng);
```
Comment (erwango) [inline drivers/entropy/entropy_stm32.c:148]: Looks it solves the on going issues and hence a welcome addition, but is there a proper explanation ?

**Comment by erwango** [comment]
> Would you be able to retest with the current version?

All tests issues are now resolved

**Review by erwango — CHANGES_REQUESTED** [review]
One question and it would be nice to update commit message to have more information on the changes that were performed.
Otherwise LGTM

File: drivers/entropy/entropy_stm32.c (hunk 2 above)
Comment (niedzwiecki-dawid) [reply drivers/entropy/entropy_stm32.c:148]: I did some testing and it looks like the IE bit in CR register is sometimes cleared when the device clock is disabled. It looks like it happens only on H7 chip. Unfortunately, I haven't fine any confirmation of such behavior in RM.

**Comment by Desvauxm-st** [comment]
LGTM for me

**Review by Desvauxm-st — APPROVED** [review]
OK for me

**Review by erwango — APPROVED** [review]

**Review by GeorgeCGV — APPROVED** [review]
Tested on H730, the `sys_rand32_get()` no longer fails.

**Comment by niedzwiecki-dawid** [comment]
Thanks for all testing and sorry for the regression.

**Pull request merged**